File: sdk/node/src/profileimport.ts
// Generic profile importer: turn a DIRECTORY of clearcote-profile JSON into a selectable index.
//
// WE ARE AN IMPORTER, NOT A HOST. Clearcote ships no third-party fingerprints and redistributes
// none. A source is just a folder of clearcote-profile JSON on the user's own machine, however
// it got there:
//
//   * the open chrome-fingerprints dataset (Vinyzu), converted locally with
//     tools/fingerprint-collect/convert_dataset.py — the dataset is fetched by the USER from the
//     original repository, so nothing is re-published here and its terms are unaffected;
//   * your own captures from tools/fingerprint-collect (collect.html / snippet.js);
//   * the licensed clearcote profile service (see profilesource.ts);
//   * any other producer that emits the same schema.
//
// Keeping the format as the ONLY contract is what makes this generic. There is no adapter per
// vendor, no vendored data, and no decompressor for someone else's archive format baked into
// three SDKs — a directory of JSON is something every source can produce and every SDK can read.

import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { gzipSync } from "node:zlib";
import { gpuVendorClass, type ProfileIndexEntry } from "./profilelib.js";

/** Parse a clearcote-profile blob into an index entry. Returns null when the file is not one. */
export function indexEntryFromProfile(
  id: string,
  profile: Record<string, unknown>,
): ProfileIndexEntry | null {
  const nav = (profile.navigator ?? {}) as Record<string, unknown>;
  const screen = (profile.screen ?? {}) as Record<string, number>;
  const webgl = (profile.webgl ?? {}) as Record<string, unknown>;
  const gl1 = (webgl.webgl1 ?? {}) as Record<string, unknown>;
  const debug = (gl1.debug ?? {}) as Record<string, string>;
  const meta = (profile.meta ?? {}) as Record<string, unknown>;

  const ua = String(nav.user_agent ?? "");
  const platform = String(nav.platform ?? "");

  // OS family from navigator.platform first (stable across UA reduction), UA as a fallback.
  let os: string;
  if (/win/i.test(platform) || /Windows/i.test(ua)) os = "windows";
  else if (/mac/i.test(platform) || /Mac OS X/i.test(ua)) os = "macos";
  else if (/linux|x11/i.test(platform) || /Linux/i.test(ua)) os = "linux";
  else if (/android/i.test(ua)) os = "android";
  else return null; // unclassifiable: better to skip than to mislabel and mis-serve it

  // Version is OPTIONAL by design. Hardware-only imports (e.g. converted chrome-fingerprints
  // records) carry no browser version and inherit the engine's — see profilelib's browser_major.
  const m = ua.match(/Chrome\/(\d+)/);
  const browserMajor = m ? Number(m[1]) : null;

  const renderer = debug.UNMASKED_RENDERER_WEBGL ?? null;

  // The launch cost, measured the same way fingerprint.ts encodes it — so an entry that would
  // blow the argv limit is excluded at selection time rather than failing at spawn.
  const encodedSize = gzipSync(Buffer.from(JSON.stringify(profile)), { level: 9 }).toString("base64").length;

  return {
    id,
    os_family: os,
    browser_major: browserMajor,
    gpu_vendor: gpuVendorClass(renderer),
    renderer: renderer ?? undefined,
    screen_width: screen.width,
    screen_height: screen.height,
    device_pixel_ratio: screen.device_pixel_ratio,
    hardware_concurrency: profile.hardware_concurrency as number | undefined,
    device_memory: profile.device_memory as number | undefined,
    encoded_size: encodedSize,
    captured_at: typeof meta.captured_at === "string" ? meta.captured_at : undefined,
  };
}

export interface ImportResult {
  index: ProfileIndexEntry[];
  /** Files that were not usable, with the reason — surfaced rather than silently dropped, so a
   *  half-converted directory is visible instead of quietly producing a thin pool. */
  skipped: { file: string; reason: string }[];
}

/**
 * Index every `*.json` in a directory.
 *
 * Reads each file once. For 10k profiles this is a few seconds and a lot of I/O, so callers
 * should cache the result (`writeIndex`) rather than re-scanning per launch.
 */
export function importDirectory(dir: string): ImportResult {
  if (!existsSync(dir) || !statSync(dir).isDirectory()) {
    throw new Error(`profile source is not a directory: ${dir}`);
  }
  const index: ProfileIndexEntry[] = [];
  const skipped: { file: string; reason: string }[] = [];

  for (const f of readdirSync(dir)) {
    if (!f.endsWith(".json")) continue;
    const path = join(dir, f);
    let parsed: Record<string, unknown>;
    try {
      parsed = JSON.parse(readFileSync(path, "utf8")) as Record<string, unknown>;
    } catch (e) {
      skipped.push({ file: f, reason: `unreadable JSON: ${String(e).slice(0, 80)}` });
      continue;
    }
    // A clearcote-profile always has navigator; without it this is some other kind of file.
    if (!parsed || typeof parsed !== "object" || !parsed.navigator) {
      skipped.push({ file: f, reason: "not a clearcote-profile (no navigator)" });
      continue;
    }
    const entry = indexEntryFromProfile(f.replace(/\.json$/, ""), parsed);
    if (!entry) {
      skipped.push({ file: f, reason: "could not classify os_family" });
      continue;
    }
    index.push(entry);
  }
  return { index, skipped };
}

/** Load one profile from an imported directory, by the id `importDirectory` assigned it. */
export function loadImportedProfile(dir: string, id: string): Record<string, unknown> {
  // The id is a filename stem this module produced, but it can reach here from a cached index,
  // so treat it as untrusted: reject anything that could escape the source directory.
  if (id.includes("/") || id.includes("\\") || id.includes("..")) {
    throw new Error(`invalid profile id '${id}'`);
  }
  const path = join(dir, `${id}.json`);
  if (!existsSync(path)) throw new Error(`profile '${id}' not found in ${dir}`);
  return JSON.parse(readFileSync(path, "utf8")) as Record<string, unknown>;
}

File: sdk/python/clearcote/_profileimport.py
"""Generic profile importer: turn a DIRECTORY of clearcote-profile JSON into a selectable index.

Port of the Node SDK's ``profileimport.ts``.

WE ARE AN IMPORTER, NOT A HOST. Clearcote ships no third-party fingerprints and redistributes
none. A source is just a folder of clearcote-profile JSON on the user's own machine, however it
got there:

* the open chrome-fingerprints dataset, converted locally with
  ``tools/fingerprint-collect/convert_dataset.py`` — fetched by the USER from the original
  repository under its own terms, so nothing is re-published here;
* your own captures from ``tools/fingerprint-collect``;
* the licensed clearcote profile service (see :mod:`_profilesource`);
* any other producer emitting the same schema.

Keeping the FORMAT as the only contract is what makes this generic: no per-vendor adapters, no
vendored data, and no third-party archive decoder baked into three SDKs.
"""

from __future__ import annotations

import base64
import gzip
import json
import os
import re
from typing import Any, Optional

from ._profilelib import gpu_vendor_class


def index_entry_from_profile(pid: str, profile: dict[str, Any]) -> Optional[dict[str, Any]]:
    """Parse a clearcote-profile blob into an index entry, or None if it is not one."""
    nav = profile.get("navigator") or {}
    screen = profile.get("screen") or {}
    webgl = profile.get("webgl") or {}
    gl1 = webgl.get("webgl1") or {}
    debug = gl1.get("debug") or {}
    meta = profile.get("meta") or {}

    ua = str(nav.get("user_agent") or "")
    platform = str(nav.get("platform") or "")

    # navigator.platform first (stable across UA reduction), UA as the fallback.
    if re.search(r"win", platform, re.I) or re.search(r"Windows", ua, re.I):
        os_family = "windows"
    elif re.search(r"mac", platform, re.I) or re.search(r"Mac OS X", ua, re.I):
        os_family = "macos"
    elif re.search(r"linux|x11", platform, re.I) or re.search(r"Linux", ua, re.I):
        os_family = "linux"
    elif re.search(r"android", ua, re.I):
        os_family = "android"
    else:
        return None  # unclassifiable: skip rather than mislabel and mis-serve it

    # Version is OPTIONAL by design. Hardware-only imports (converted chrome-fingerprints records)
    # carry no browser version and inherit the engine's — see _profilelib.eligible.
    m = re.search(r"Chrome/(\d+)", ua)
    browser_major = int(m.group(1)) if m else None

    renderer = debug.get("UNMASKED_RENDERER_WEBGL")

    # The real launch cost, encoded exactly as _fingerprint.py encodes it, so an entry that would
    # blow the argv limit is excluded at SELECTION time rather than failing at spawn.
    raw = json.dumps(profile, separators=(",", ":")).encode("utf-8")
    encoded_size = len(base64.b64encode(gzip.compress(raw, 9)))

    return {
        "id": pid,
        "os_family": os_family,
        "browser_major": browser_major,
        "gpu_vendor": gpu_vendor_class(renderer),
        "renderer": renderer,
        "screen_width": screen.get("width"),
        "screen_height": screen.get("height"),
        "device_pixel_ratio": screen.get("device_pixel_ratio"),
        "hardware_concurrency": profile.get("hardware_concurrency"),
        "device_memory": profile.get("device_memory"),
        "encoded_size": encoded_size,
        "captured_at": meta.get("captured_at") if isinstance(meta.get("captured_at"), str) else None,
    }


def import_directory(directory: str) -> dict[str, Any]:
    """Index every ``*.json`` in a directory.

    Returns ``{"index": [...], "skipped": [{"file":…, "reason":…}]}``. Skips are REPORTED rather
    than swallowed, so a half-converted directory is visible instead of silently producing a
    thin pool.
    """
    if not os.path.isdir(directory):
        raise ValueError(f"profile source is not a directory: {directory}")

    index: list[dict[str, Any]] = []
    skipped: list[dict[str, str]] = []

    for name in sorted(os.listdir(directory)):
        if not name.endswith(".json"):
            continue
        path = os.path.join(directory, name)
        try:
            with open(path, "r", encoding="utf-8") as f:
                parsed = json.load(f)
        except (OSError, ValueError) as e:
            skipped.append({"file": name, "reason": f"unreadable JSON: {str(e)[:80]}"})
            continue
        # A clearcote-profile always has navigator; without it this is some other kind of file.
        if not isinstance(parsed, dict) or not parsed.get("navigator"):
            skipped.append({"file": name, "reason": "not a clearcote-profile (no navigator)"})
            continue
        entry = index_entry_from_profile(name[: -len(".json")], parsed)
        if entry is None:
            skipped.append({"file": name, "reason": "could not classify os_family"})
            continue
        index.append(entry)

    return {"index": index, "skipped": skipped}


def load_imported_profile(directory: str, pid: str) -> dict[str, Any]:
    """Load one profile from an imported directory, by the id ``import_directory`` assigned."""
    # The id is a filename stem this module produced, but it can arrive here from a cached index,
    # so treat it as untrusted: reject anything that could escape the source directory.
    if "/" in pid or "\\" in pid or ".." in pid:
        raise ValueError(f"invalid profile id '{pid}'")
    path = os.path.join(directory, f"{pid}.json")
    if not os.path.isfile(path):
        raise ValueError(f"profile '{pid}' not found in {directory}")
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)
